### PR TITLE
Merge unit tests

### PR DESCRIPTION
**Pull request by Jodasue**
The following branch has three commits.

1. Unit tests
2. Bug fixes to exarl related to the unit tests
3. Refactor workflows and updates to the unit test docs

The refactors workflows include simple and simple_async.  This is an update software architecture that allows us to build more robust workflows which are more easily tested.

### CODE REVIEW

File: exarl/agents/__init__.py [hunk 1]
```diff
@@ -1,9 +1,6 @@
 from exarl.agents.registration import register, make
 import exarl.utils.candleDriver as cd
-try:
-    agent = cd.run_params['agent']
-except:
-    print("Unknown agent: {} did not match any registered agents.".format(agent))
+agent = cd.lookup_params('agent')
```
Comment (rvinaybharadwaj) [inline exarl/agents/__init__.py:3]: @jmohdyusof will let you comment on this.

File: exarl/base/agent_base.py [hunk 2]
```diff
@@ -52,37 +54,56 @@ def get_weights(self):
...
     @abstractmethod
-    def train(self):
+    def train(self, batch):
         """train the agent
         """
         pass
 
     @abstractmethod
-    def action(self):
+    def target_train(self):
```
Comment (rvinaybharadwaj) [inline exarl/base/agent_base.py:69]: I am not sure if we should mandate target_train as some algorithms use a single active policy. SMEs should comment on this.

**Review by jmohdyusof — DISMISSED** [review]
We removed lookup_params some time ago, because it causes circular imports and adds confusion about where parameters are being set. The correct way is to set values in the config file or via command line, and then read the dictionary item, as it is done in develop. 

Please remove all calls to lookup_params.

**Review by rvinaybharadwaj — CHANGES_REQUESTED** [review]

**Comment by Jodasue** [comment]
I tried to work without (manually removed all of them) on Saturday.
Unfortunately I have issues with the unit tests.  I am happy to go over the
problem with you later today or tomorrow.
-Josh S.

**Comment by rvinaybharadwaj** [comment]
I believe the docs build error can be fixed by specifying to exclude the hardrec submodule: https://github.com/exalearn/EXARL/blob/43a613a6472c11f582b137cbae3468b38760acb7/docs/source/conf.py#L74

**Comment by rvinaybharadwaj** [comment]
I ran both the old and new workflows and here are the results:
CartPole using async (2 processes, 100 episodes, 100 steps/episode)
CP_async
CartPole using simple_async (2 processes, 100 episodes, 100 steps/episode)
CP_simple_async

I'm checking to see if the stopping criteria is different in simple_async. Also, the reward is consistently lower in simple_async compared to sync. I'm checking to see if the default parameters are different in any way.

File: exarl/workflows/workflow_vault/simple_learner.py [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+            if policy_type == 0:
+                workflow.agent.epsilon_adj()
+
+            if done:
+                self.done_episode += 1
+                self.episode_per_rank[src] = self.next_episode
+                self.next_episode += 1
+
+        for dst in to_send:
+            self.send_model(workflow, self.episode_per_rank[dst], self.train_return[dst], dst)
```
Comment (rvinaybharadwaj) [inline exarl/workflows/workflow_vault/simple_learner.py:454]: Please replace workflow with exalearner

**Review by rvinaybharadwaj — CHANGES_REQUESTED** [review]

File: exarl/workflows/workflow_vault/simple_learner.py [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+        ----------
+        workflow : ExaWorkflow
+            This contains the agent and env
+        """
+        nepisodes = workflow.nepisodes
+        if ExaComm.global_comm.rank == 0:
+            if self.block_size == ExaComm.agent_comm.size and ExaComm.agent_comm.size > 1:
+                nactors = ExaComm.global_comm.size - ExaComm.num_learners
+                if workflow.nepisodes % nactors:
+                    nepisodes = (int(workflow.nepisodes / nactors) + 1) * nactors
```
Comment (rvinaybharadwaj) [inline exarl/workflows/workflow_vault/simple_learner.py:570]: I don't quite understand how this works when procs_per_env > 1

Comment (rvinaybharadwaj) [reply exarl/workflows/workflow_vault/simple_learner.py:570]: nactors = ExaComm.agent_comm.size - ExaComm.num_learners

**Comment by Jodasue** [comment]
> I ran both the old and new workflows and here are the results
> 
> I'm checking to see if the stopping criteria is different in simple_async. Also, the reward is consistently lower in simple_async compared to sync. I'm checking to see if the default parameters are different in any way.

We looked into this and found it was epsilon that was not being updated.

File: utests/__init__.py [hunk 5]
```diff
@@ -1 +1,8 @@
-from exarl.agents.agent_vault.dqn import DQN
+from exarl.utils import candleDriver
+try:
+    
+    # candleDriver.initialize_parameters()
+    print("Loaded this here")
```
Comment (rvinaybharadwaj) [inline utests/__init__.py:5]: Are we missing something here?

**Review by rvinaybharadwaj — CHANGES_REQUESTED** [review]

**Comment by rvinaybharadwaj** [comment]
I encounter a bug when I try to run 

```
Traceback (most recent call last):
  File "exarl/driver/__main__.py", line 32, in <module>
    exa_learner = exarl.ExaLearner()
  File "/projects/users/vinayr/ExaLearn/ExaRL/exarl/base/learner_base.py", line 40, in __init__
    ExaSimple(comm, process_per_env, learner_procs)
  File "/projects/users/vinayr/ExaLearn/ExaRL/exarl/network/simple_comm.py", line 63, in __init__
    super().__init__(self, procs_per_env, num_learners)
  File "/projects/users/vinayr/ExaLearn/ExaRL/exarl/base/comm_base.py", line 26, in __init__
    ExaComm.agent_comm, ExaComm.env_comm, ExaComm.learner_comm = comm.split(procs_per_env, num_learners)
  File "/projects/users/vinayr/ExaLearn/ExaRL/exarl/network/simple_comm.py", line 202, in split
    agent_comm = ExaSimple(comm=agent_comm)
TypeError: __init__() missing 2 required positional arguments: 'procs_per_env' and 'num_learners'
```

File: utests/__init__.py (hunk 5 above)
Comment (Jodasue) [reply utests/__init__.py:5]: Also on it.

File: exarl/utils/globals.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        """
+        if name not in ExaGlobals.__logger:
+            if ExaGlobals.is_init():
+                ExaGlobals.__logger[name] = ExaGlobals.__init_log(name, level)
+            else:
+                ExaGlobals.__logger[name] = level
+        return lambda : ExaGlobals.log_helper(name)
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (Jodasue) [reply exarl/utils/globals.py:189]: Done.

File: exarl/agents/agent_vault/ddpg.py [hunk 7]
```diff
@@ -66,28 +58,28 @@ def __init__(self, env, is_learner):
         self.upper_bound = env.action_space.high
         self.lower_bound = env.action_space.low
 
-        logger.info("Size of State Space:  {}".format(self.num_states))
-        logger.info("Size of Action Space:  {}".format(self.num_actions))
-        logger.info('Env upper bounds: {}'.format(self.upper_bound))
-        logger.info('Env lower bounds: {}'.format(self.lower_bound))
+        logger().info("Size of State Space:  {}".format(self.num_states))
```
[the remark this answers is not in the document]
Comment (Jodasue) [reply exarl/agents/agent_vault/ddpg.py:61]: Yes, it was the best I could come up with.  I needed to delay the creation of the logger until after we read in the configuration files.  This is now part of globals and the lambda allows for the delay.

File: exarl/config/workflow_cfg/simple.json [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{
```
[the remark this answers is not in the document]
Comment (Jodasue) [reply exarl/config/workflow_cfg/simple.json:1]: I would like to have a commit that has both and then I will do a separate commit with just the overwrite.  That way if we need to go back in the history we can make sense of what is going on.

File: exarl/driver/__main__.py [hunk 9]
```diff
@@ -36,7 +36,7 @@
 rank = comm.rank
 size = comm.size
 
-writeDir = cd.run_params["introspector_dir"]
+writeDir = ExaGlobals.lookup_params("introspector_dir")
 if writeDir != "none":
```
[the remark this answers is not in the document]
Comment (Jodasue) [reply exarl/driver/__main__.py:40]: Fixed.

File: utests/__init__.py (hunk 5 above)
Comment (Jodasue) [reply utests/__init__.py:5]: Cleaned this up, but this file is empty now on purpose.

**Review by rvinaybharadwaj — APPROVED** [review]

**Review by rvinaybharadwaj — APPROVED** [review]

**Comment by jmohdyusof** [comment]
Passes required tests

**Review by jmohdyusof — APPROVED** [review]

**Pull request merged**